### DISCUSSION TITLE
Option for escaping dropped paths for xonsh

### DISCUSSION THREAD

**Discussion by CouleeApps** [ghostty-org/ghostty#3548] [Feature Requests, Ideas]
When dragging and dropping files into Ghostty, their paths are escaped by adding backslashes before separator characters. However, users of xonsh [need a different escaping strategy](https://xon.sh/tutorial_subproc_strings.html#no-escape) since backslashes are not supported. I know of two different ways this is addressed in existing tools:

* In Terminal.app, you can hold Shift when dropping a file and the name will not be escaped at all. You get something like `glennsmith@Glenns-MacBook-Pro ~ @ /Applications/Binary Ninja.app` Technically this forms an invalid command, but you can pretty easily muscle memory typing an open quote, dragging in the file, and typing the closing quote.
* In iTerm2, there is an experimental option which replaces the backslash escaping with automatically adding single quotes around the file name. You get something like `glennsmith@Glenns-MacBook-Pro ~ @ '/Applications/Binary Ninja.app'` though actually iTerm's implementation only adds quotes if necessary, which makes it rather finicky when dropping files since you can never be sure if they will become a string or will just be pasted raw.

So while there are tradeoffs to both solutions, it is practically necessary to implement at least one of them before ghostty is possible to use with xonsh.

**Comment by oderwat** [comment]
This also is true for NuShell:

As a NuShell user i run into the problem that drag & drop of a file - for example from Finder - into Ghostty will result in something like `/Users/anon/text\ file \(1\).txt`.

This is not supported with NuShell and it would be very helpfull if such a drop action would result in `"/Users/anon/text file (1).txt"` instead.

I think this could be helpful in other cases too. For example when pasting into Helix I also get the same escapes.

I imagine that one could define the default paste mode between "escaped" "double-quotes" "single-quotes" "raw"?

Maybe it is also possible to have the mode change with different key modifiers, like "shift drop" means "quote escaped".

**Comment by kLy** [comment]
No movement on this front? This issue is really bad for nushell as it makes it impossible to navigate to a folder quickly.

Combined with #11142, things get really aggravated, as there is literally zero workarounds to be able to quickly navigate to a folder.

Could we get this prioritised? Nushell and xonsh are both in the top 10 list of most used unix shells and currently ghostty is very broken for both.

**Reply by pluiedev** [reply]
We are all volunteers and nobody is obligated to fix anything in particular. We will get to this in due time, or you can try to fix this yourself before us. Just sitting here and complaining about it does nothing.

**Comment by anki-code** [comment]
In xonsh shell you can use `BracketedPaste` and process the pasted text (folks from Nu can copy this approach):

https://github.com/user-attachments/assets/0c8be916-cb4e-4f79-8797-1b3651aec1ef

Use [xonsh RC](https://xon.sh/xonshrc.html):
```xsh
@events.on_ptk_create
def _bracketed_paste_path_fix(bindings, **kw):
    @bindings.add(@.imp.prompt_toolkit.keys.Keys.BracketedPaste)
    def _paste(event):
        """If line looks like escaped path transform it to quoted path."""
        import shlex
        data = event.data
        data = data.replace("\r\n", "\n").replace("\r", "\n")

        looks_like_drop = (
            "\n" not in data
            and "\\" in data
            and ("/" in data or data.startswith("~"))
        )

        if looks_like_drop:
            try:
                parts = shlex.split(data, posix=True)
                data = " ".join(shlex.quote(p) for p in parts)
            except ValueError:
                pass
        event.current_buffer.insert_text(data)

```

Or create your own binding:

<details>

```xsh
# ~/.xonshrc
from prompt_toolkit.keys import Keys
from prompt_toolkit.application import get_app

@events.on_ptk_create
def custom_keybindings(bindings, **kw):

    @bindings.add("c-v")  # Ctrl+V for pasting
    def paste_preprocessed(event):
        app = get_app()
        clipboard = app.clipboard.get_data()
        pasted_text = clipboard.text

        # Convert to uppercase and paste.
        processed_text = pasted_text.upper()
        event.app.current_buffer.insert_text(processed_text)
```
If you find this useful and create a solution feel free to create a [xontrib](https://github.com/topics/xontrib) using [xontrib-template](https://github.com/xonsh/xontrib-template).

</details>
